sky/spot/recovery_strategy.py

```diff
@@ -142,39 +156,66 @@ def _launch(self, max_retry=3, raise_on_failure=True) -> Optional[float]:
+                    retry_launch = True
+                    logger.info(
+                        'Failed to get the job status, due to unexpected '
+                        'job submission error.')
+                    break
+                cluster_status, _ = backend_utils.refresh_cluster_status_handle(
+                    self.cluster_name, force_refresh=True)
+                if cluster_status != global_user_state.ClusterStatus.UP:
+                    # The cluster can be preempted before the job is launched.
+                    # In this case, we will terminate the cluster and retry
+                    # the launch.
+                    retry_launch = True
+                    logger.info(
+                        'The cluster is preempted before the job starts.')
+                    break
+                # TODO(zhwu): The preemption can still happen right before this
+                # line, but it is very unlikely.
                 status = spot_utils.get_job_status(self.backend,
                                                    self.cluster_name)
-                while (status is None or status in [
-                        job_lib.JobStatus.INIT,
-                        job_lib.JobStatus.PENDING,
-                ]):
-                    time.sleep(spot_utils.JOB_STARTED_STATUS_CHECK_GAP_SECONDS)
-                    status = spot_utils.get_job_status(self.backend,
-                                                       self.cluster_name)
-                launch_time = spot_utils.get_job_timestamp(self.backend,
-                                                           self.cluster_name,
-                                                           get_end_time=False)
-                return launch_time
-
-            if max_retry is not None and retry_cnt >= max_retry:
-                # Retry forever if max_retry is None.
-                if raise_on_failure:
-                    with ux_utils.print_exception_no_traceback():
-                        raise exceptions.ResourcesUnavailableError(
-                            'Failed to launch the spot cluster after '
-                            f'{max_retry} retries.')
-                else:
-                    return None
-            gap_seconds = backoff.current_backoff()
-            logger.info(
-                f'Retrying to launch the spot cluster in {gap_seconds:.1f} '
-                'seconds.')
-            time.sleep(gap_seconds)
+                if status not in [
+                        None, job_lib.JobStatus.INIT, job_lib.JobStatus.PENDING
+                ]:
+                    # The job has been transited into a non-initial status. We
+                    # can continue the logic.
+                    break
+                # Wait the job to be started
+                time.sleep(spot_utils.JOB_STARTED_STATUS_CHECK_GAP_SECONDS)
+
+            if retry_launch:
+                self._cancel_all_jobs()
```

Comment:
`self._try_cancel_all_jobs()`?  Directly indicates any exception (e.g., cluster is down) would be ok.

Reply:
Good point! Done. Thanks!